File: Sources/HaulCore/Core/Log.swift
import Foundation
import Synchronization
#if canImport(Glibc)
import Glibc
#elseif canImport(Darwin)
import Darwin
#endif

/// Console output. Normal mode prints clean status lines with a leading glyph; `--debug` adds timestamps
/// and shows the internal chatter. Colour follows the terminal (`NO_COLOR` respected).
///
/// Everything goes to stdout, unless `toStderr` is set (`--json`): then stdout carries only the JSON document.
public enum Log {
    private static let debugFlag = Atomic<Bool>(false)
    private static let stderrFlag = Atomic<Bool>(false)
    public static var style: Style { Style(color: Terminal.supportsColor) }

    public static var debugEnabled: Bool {
        get { debugFlag.load(ordering: .relaxed) }
        set { debugFlag.store(newValue, ordering: .relaxed) }
    }

    /// Send every log line to stderr, keeping stdout for data.
    public static var toStderr: Bool {
        get { stderrFlag.load(ordering: .relaxed) }
        set { stderrFlag.store(newValue, ordering: .relaxed) }
    }

    /// The file descriptor log lines go to.
    public static var fd: Int32 { toStderr ? STDERR_FILENO : STDOUT_FILENO }

    public static var isTerminal: Bool { Terminal.isTTY }

    static func write(_ s: String) {
        if toStderr {
            FileHandle.standardError.write(Data(s.utf8))
        } else {
            print(s, terminator: "")
            fflush(stdout)
        }
    }

    /// Data for stdout, whatever `toStderr` says.
    public static func output(_ s: String) {
        print(s)
        fflush(stdout)
    }

    public static func timestamp(_ date: Date = Date()) -> String {
        let c = Calendar.current.dateComponents([.year, .month, .day, .hour, .minute, .second, .nanosecond], from: date)
        let ms = (c.nanosecond ?? 0) / 1_000_000
        return String(format: "[%04d-%02d-%02d %02d:%02d:%02d.%03d]",
                      c.year ?? 0, c.month ?? 0, c.day ?? 0, c.hour ?? 0, c.minute ?? 0, c.second ?? 0, ms)
    }

    private static func emit(_ glyph: String, _ text: String, newline: Bool = true) {
        let prefix = debugEnabled ? style.gray(timestamp()) + " " : ""
        write(prefix + glyph + text + (newline ? "\n" : ""))
    }

    /// A neutral line.
    public static func info(_ text: String, newline: Bool = true) { emit("  ", text, newline: newline) }
    /// Low-key activity (loading a cookie, a subtitle found…).
    public static func status(_ text: String) { emit("  ", style.dim("· " + text)) }
    public static func success(_ text: String) { emit(style.green("✓ "), text) }
    public static func warn(_ text: String) { emit(style.yellow("! "), style.yellow(text)) }
    public static func error(_ text: String) { emit(style.red("✗ "), style.red(text)) }
    public static func highlight(_ text: String) { emit("  ", style.cyan(text)) }
    public static func title(_ text: String) { emit("  ", style.bold(text)) }
    public static func banner(_ text: String) { write(style.bold(text) + "\n") }
    public static func plain(_ text: String = "") { write(text + "\n") }
    public static func lines(_ lines: [String]) { for l in lines { plain(l) } }
    /// A question; the cursor stays on the line.
    public static func prompt(_ text: String) { emit("  ", style.cyan(text), newline: false) }

    public static func debug(_ text: @autoclosure () -> String) {
        guard debugEnabled else { return }
        write(style.gray(timestamp() + " " + text()) + "\n")
    }
}

File: internal/console/log.go
package console

import (
	"fmt"
	"os"
	"sync"
	"sync/atomic"
	"time"
)

var (
	debug    atomic.Bool
	toStderr atomic.Bool
	writeMu  sync.Mutex
)

// Debug reports whether debug output (--debug) is on.
func Debug() bool { return debug.Load() }

// SetDebug turns debug output on or off.
func SetDebug(on bool) { debug.Store(on) }

// ToStderr reports whether log lines go to stderr.
func ToStderr() bool { return toStderr.Load() }

// SetToStderr sends every log line to stderr, keeping stdout for data.
func SetToStderr(on bool) { toStderr.Store(on) }

// Write puts raw text on the log stream.
func Write(s string) {
	writeMu.Lock()
	defer writeMu.Unlock()
	_, _ = Out().WriteString(s)
}

// Output writes data (the --json document) to stdout, whatever ToStderr says.
func Output(s string) {
	writeMu.Lock()
	defer writeMu.Unlock()
	_, _ = os.Stdout.WriteString(s + "\n")
}

// Timestamp is [2006-01-02 15:04:05.000].
func Timestamp(t time.Time) string { return t.Format("[2006-01-02 15:04:05.000]") }

func emit(glyph, text string, newline bool) {
	prefix := ""
	if Debug() {
		prefix = CurrentStyle().Gray(Timestamp(time.Now())) + " "
	}
	end := ""
	if newline {
		end = "\n"
	}
	Write(prefix + glyph + text + end)
}

// Info is a neutral line.
func Info(text string) { emit("  ", text, true) }

// Status is low-key activity: a cookie loaded, a subtitle found…
func Status(text string) { emit("  ", CurrentStyle().Dim("· "+text), true) }

// Success is a ✓ line.
func Success(text string) { emit(CurrentStyle().Green("✓ "), text, true) }

// Warn is a ! line for a problem that does not stop the run.
func Warn(text string) { s := CurrentStyle(); emit(s.Yellow("! "), s.Yellow(text), true) }

// Error is a ✗ line.
func Error(text string) { s := CurrentStyle(); emit(s.Red("✗ "), s.Red(text), true) }

// Highlight is a cyan line.
func Highlight(text string) { emit("  ", CurrentStyle().Cyan(text), true) }

// Banner is a bold line without indent.
func Banner(text string) { Write(CurrentStyle().Bold(text) + "\n") }

// Plain is a line as it is.
func Plain(text string) { Write(text + "\n") }

// Lines prints each line.
func Lines(lines []string) {
	for _, l := range lines {
		Plain(l)
	}
}

// Prompt is a question; the cursor stays on the line.
func Prompt(text string) { emit("  ", CurrentStyle().Cyan(text), false) }

// Debugf prints only under --debug, with a timestamp.
func Debugf(format string, a ...any) {
	if !Debug() {
		return
	}
	Write(CurrentStyle().Gray(Timestamp(time.Now())+" "+fmt.Sprintf(format, a...)) + "\n")
}
